### PR TITLE
172411943 measurement class

### PR DESCRIPTION
**Pull request by simonwa7**
Creating a class for `Measurements` and reconciling overlap with `BitstringDistribution` class. 

This PR also affects the `QuantumSimulator` interface and therefore resources implementing this interface may need to be updated

### CODE REVIEW

**Comment by simonwa7** [comment]
This PR also fixes versioning of packages that have had breaking changes (`cirq` and `qiskit`)

File: src/python/zquantum/core/interfaces/backend.py [hunk 1]
```diff
@@ -23,6 +23,8 @@ def run_circuit_and_measure(self, circuit:Circuit, **kwargs) -> List[Tuple]:
         Method for executing the circuit and measuring the outcome.
         Args:
             circuit (core.circuit.Circuit): quantum circuit to be executed.
+        Returns:
+            Measurements: object representing the measurements resulting from the circuit
```
Comment (mstechly) [inline src/python/zquantum/core/interfaces/backend.py:27]: Actually it should be named "core.measurements.Measurements" I guess ;)

File: src/python/zquantum/core/measurement.py [hunk 2]
```diff
@@ -392,3 +393,115 @@ def convert_bitstring_to_int(bitstring: Iterable[int]) -> int:
...
+            filename (string): filename to save the data to 
+        """
+        data = { "schema":         SCHEMA_VERSION+"-measurements",
+                 "counts":         self.get_counts(),
+                 "measurements":   self.get_measurements()}
+        with open(filename, "w") as f:
+            f.write(json.dumps(data, indent=2))
+
+
+    def get_measurements(self):
```
Comment (mstechly) [inline src/python/zquantum/core/measurement.py:424]: Why do you need a getter for that?
The fields in the method are not private, so you can just do:
`measurements.measurements`, not need for `measurements.get_measurements()`.
You now have two ways to do the same thing.

File: src/python/zquantum/core/measurement.py [hunk 3]
```diff
@@ -392,3 +393,115 @@ def convert_bitstring_to_int(bitstring: Iterable[int]) -> int:
             significant (little endian).
     """
     return int("".join(str(bit) for bit in bitstring[::-1]), 2)
+
+class Measurements:
+    """ A class representing measurements from a quantum circuit. """
+
+    def __init__(self, filename=None):
+        self.counts = {}
+        self.measurements = []
```
Comment (mstechly) [inline src/python/zquantum/core/measurement.py:402]: Actually, perhaps "bitstrings" is better?
Cause measurements.measurements looks a little bit confusing to me ;)

File: src/python/zquantum/core/measurement.py [hunk 4]
```diff
@@ -392,3 +393,115 @@ def convert_bitstring_to_int(bitstring: Iterable[int]) -> int:
...
+    def add_measurements(self, measurements):
+        """ Add a list of measurement tuples 
+
+        Args:
+            measurements (list of tuples): A list of the measurements to add
+        """
+        for measurement in measurements:
+            self.measurements.append(tuple(measurement))
+
+            # Bitstrings are in reversed-index order compared to measurement tuples
```
Comment (mstechly) [inline src/python/zquantum/core/measurement.py:460]: Why?
I don't think we should have such inconsistencies in the core library.

File: src/python/zquantum/core/measurement.py [hunk 5]
```diff
@@ -392,3 +393,115 @@ def convert_bitstring_to_int(bitstring: Iterable[int]) -> int:
...
+        return self.num_measurements
+
+
+    def add_measurements(self, measurements):
+        """ Add a list of measurement tuples 
+
+        Args:
+            measurements (list of tuples): A list of the measurements to add
+        """
+        for measurement in measurements:
```
Comment (mstechly) [inline src/python/zquantum/core/measurement.py:457]: This for loop (and the following ones as well) look like potential performance bottleneck.
Isn't there a more pythonic way to do that?

File: src/python/zquantum/core/measurement.py [hunk 6]
```diff
@@ -392,3 +393,115 @@ def convert_bitstring_to_int(bitstring: Iterable[int]) -> int:
...
+                bitstring += str(bit_value)
+            
+            if bitstring in self.counts.keys():
+                self.counts[bitstring] += 1
+            else:
+                self.counts[bitstring] = 1
+            self.num_measurements += 1
+
+
+    def add_counts(self, counts):
```
Comment (mstechly) [inline src/python/zquantum/core/measurement.py:472]: Why do we need to pass `counts` as an argument?
Shouldn't it calculate counts directly from `self.measurements`?

File: src/python/zquantum/core/interfaces/backend.py (hunk 1 above)
Comment (simonwa7) [reply src/python/zquantum/core/interfaces/backend.py]: Addressed here: 
5309e30

File: src/python/zquantum/core/measurement.py (hunk 3 above)
Comment (simonwa7) [reply src/python/zquantum/core/measurement.py]: Addressed here: 
9e00518

File: src/python/zquantum/core/measurement.py (hunk 6 above)
Comment (simonwa7) [reply src/python/zquantum/core/measurement.py:472]: This is to allow the user to add measurements to the object in either form. I believe many simulators (qiskit for example) return similar objects to the `counts` dictionary, so this avoids them having to create lists of tuples when they really just want to add the `counts`.

File: src/python/zquantum/core/measurement.py (hunk 5 above)
Comment (simonwa7) [reply src/python/zquantum/core/measurement.py]: Addressed in refactor here: 
231a59a

File: src/python/zquantum/core/measurement.py (hunk 4 above)
Comment (simonwa7) [reply src/python/zquantum/core/measurement.py]: Addressed in refactor here:
231a59a

File: src/python/zquantum/core/measurement.py (hunk 2 above)
Comment (simonwa7) [reply src/python/zquantum/core/measurement.py]: Addressed in refactor here:
231a59a

**Comment by simonwa7** [comment]
Since we chose to order measurement tuples in reverse index order, we should make sure measurements are properly indexed when we propagate these changes to the resources implementing the QuantumSimulator interface

File: src/python/zquantum/core/measurement.py [hunk 7]
```diff
@@ -392,3 +393,96 @@ def convert_bitstring_to_int(bitstring: Iterable[int]) -> int:
...
+        """ Get the measurements as a histogram
+
+        Returns:
+            A dictionary mapping bitstrings to integers representing the number of times the bitstring was measured
+        """
+        bitstrings = convert_tuples_to_bitstrings(self.bitstrings)
+        return dict(Counter(bitstrings))
+
+
+    def add_counts(self, counts: Dict):
```
Comment (mstechly) [inline src/python/zquantum/core/measurement.py:461]: FYI you an specify what should be the form of the Dict – e.g. Dict[Tuple[int], int] .
Not necessary to do so though.

File: src/python/zquantum/core/measurement.py [hunk 8]
```diff
@@ -392,3 +393,96 @@ def convert_bitstring_to_int(bitstring: Iterable[int]) -> int:
...
+        else:
+            data = json.load(file)
+
+        bitstrings = []
+        for bitstring in data["bitstrings"]:
+            bitstrings.append(tuple(bitstring))
+
+        return cls(bitstrings=bitstrings)
+
+    def save(self, filename: String):
```
Comment (mstechly) [inline src/python/zquantum/core/measurement.py:438]: I think it should be `str` not `String` – the typing requires those funny capital letter names only for the "special type" or "collection type" of variables.
See: https://docs.python.org/3/library/typing.html

File: src/python/zquantum/core/measurement.py [hunk 9]
```diff
@@ -392,3 +393,115 @@ def convert_bitstring_to_int(bitstring: Iterable[int]) -> int:
...
+
+class Measurements:
+    """ A class representing measurements from a quantum circuit. """
+
+    def __init__(self, filename=None):
+        self.counts = {}
+        self.bitstrings = []
+        self.num_measurements = 0
+
+        if filename is not None:
```
[the remark this answers is not in the document]
Comment (simonwa7) [reply src/python/zquantum/core/measurement.py]: Already addressed in previous commits

File: src/python/zquantum/core/measurement.py (hunk 7 above)
Comment (simonwa7) [reply src/python/zquantum/core/measurement.py:461]: Good to know!

File: src/python/zquantum/core/measurement.py (hunk 8 above)
Comment (simonwa7) [reply src/python/zquantum/core/measurement.py]: Addressed here: 
7288dca

**Pull request merged**